ADR/0033-product-version-controller.md

```diff
@@ -0,0 +1,493 @@
+managing the underlying Tekton Pipeline and other lower level objects.
+
+With Konflux shifting focus to serving enterprise development teams requests
+had been made to enable using other abstraction metaphors such as "Product" and
+"Version" for managing development, build, test and release workflows. As one
+user put it:
+
+> As an architect, I need to be able to see all my apps versions in
+> one place, so that they are easily accessible by everyone on the
+> team and organized in a way that matches the way our team talks
+> about them, making my workflow for each app more streamlined.
+
+Some Konflux users have already attempted managing large multi-versioned
+products in Konflux. The pattern that emerged for doing so is as follows:
+
+* Create a separate Konflux Application for each product version
+* Use branches to track different source code versions within Git repositories
+* Use a separate Konflux Components for each buildable piece of code
+  (E.g. Dockerfile) in each branch in each repo that is included in a particular
+  application version.
+* To manage large amounts of components, some users have used multiple
+  application objects for a single product version.
+
+As one can expect, creating and managing so many Konflux objects is currently
+rather cumbersome. There are a few redundant steps that need to be taken. For
+example, access credentials for a Git repository need to be filled-in every time
+a component referring to that repository is created, even if other components
+referring that repository already exist in the user's workspace.
+
+The intention for the solution proposed in this document is to follow the
+pattern described above as far as the Konflux "Application" and "Component"
+objects go and provide a higher-level layer for the streamlined management of
+large amounts of such objects.
+
+An alternative approach to managing the challenges we have described above is to
+modify the way Application and Component object function so that for example:
+
+* A component may encompass multiple versions of similar code across different
+  Git branches
+* An application may include several sub groups of components so that each could
+  be regarded to be defining a different version
+
+We see several benefits to the "high level layer" approach we propose over the
+alternative:
+
+* The way Components and Applications work is currently tightly coupled to the
+  way the various Konflux services work. Changing it would require changes
+  across multiple services and may introduce operational risks.
+* There are proposals being discussed for decoupling the various Konflux
+  services and possibly deprecating the Component and Application objects.
+  Having a separate high-level layer would allow for adjusting it for using
+  different underlying objects while continued emphasis on using Application
+  and Component object would necessitate continued maintenance of them.
```

Comment:
How would this higher level layer benefit in the decoupling process? If we have some higher level representation, we don't want it to just become _another_ API that needs to adjust/migrate during the decoupling.

Reply:
It easier to adjust a higher-level layer to changing lower-level objects - then it is to port core functionality from one set of objects to another.